crypto/sm2/sm2_sign.c

```diff
@@ -0,0 +1,193 @@
+/*
+ * Copyright 2017 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright 2017 Ribose Inc. All Rights Reserved.
+ * Ported from Ribose contributions from Botan.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/sm2.h>
+#include <openssl/evp.h>
+
+static BIGNUM *compute_msg_hash(const EVP_MD *digest,
+                                const EC_KEY *key,
+                                const char *user_id,
+                                const uint8_t *msg, size_t msg_len)
+{
+    EVP_MD_CTX *hash = EVP_MD_CTX_new();
+    const int md_size = EVP_MD_size(digest);
+    uint8_t za[md_size];
+    BIGNUM *e = NULL;
+
+    if (SM2_compute_za(za, digest, user_id, key) == 0)
+        goto done;
+
+    if (EVP_DigestInit(hash, digest) == 0)
+        goto done;
+
+    if (EVP_DigestUpdate(hash, za, md_size) == 0)
+        goto done;
+
+    if (EVP_DigestUpdate(hash, msg, msg_len) == 0)
+        goto done;
+
+    // reuse za buffer to hold H(ZA || M)
+    if (EVP_DigestFinal(hash, za, NULL) == 0)
+        goto done;
+
+    e = BN_bin2bn(za, md_size, NULL);
+
+ done:
+    EVP_MD_CTX_free(hash);
+    return e;
+}
+
+ECDSA_SIG *SM2_do_sign(const EC_KEY *key,
+                       const EVP_MD *digest,
+                       const char *user_id, const uint8_t *msg, size_t msg_len)
+{
+    ECDSA_SIG *sig = NULL;
+    BIGNUM *k = BN_new();
+    BIGNUM *rk = BN_new();
+    BIGNUM *e = NULL;
+    BIGNUM *r = BN_new();
+    BIGNUM *s = BN_new();
+    BIGNUM *x1 = BN_new();
+    BIGNUM *tmp = BN_new();
+    BN_CTX *ctx = BN_CTX_new();
```

Comment:
Again why not move the ctx declaration higher and use the ctx for some of these BIGNUM allocations, the ones that are only in scope within this function, ie. k, x1 and tmp